### ISSUE TITLE
docstore: matching error specs with mongodb

### ISSUE THREAD

**Issue by jba** [google/go-cloud#1660]
Docstore specifies the following behaviors, which seem impossible to satisfy in MongoDB with a single RPC:
- When writing a document with a revision, if the document doesn't exist a NotFound error is returned, but if it exists with the wrong revision then a FailedPrecondition error is returned.
- When deleting a document with a revision, if the document doesn't exist no error is returned, but if it exists with the wrong revision then a FailedPrecondition error is returned.

Mongo doesn't support revisions natively, so the client will manage them itself, by reading and writing the "DocstoreRevision" field of documents. To write or delete a document with a given ID `id` at revision `rev`, we select for the document using the "filter" `{"_id": id, "DocstoreRevision": rev}`. If there is a matching document, this works as intended: the document is written or deleted. But if no document matches, it could be for two reasons: either the document doesn't exist (no matching ID) or it exists at the wrong revision. Mongo doesn't distinguish between these two: it just tells us that there were no matches to our filter. 

We could distinguish the two cases by first looking up the document just by ID. But making two RPCs for every write and delete (that involves revisions) would be unacceptable for performance.

Trying to do the lookup in parallel won't work in general. If we're deleting a document, the lookup would race with the delete. It might work for writes, though.

We could loosen the semantics of writes, saying that a write with a bad revision might return either FailedPrecondition or NotFound. That doesn't help with deletes, though.

We could change the semantics of Delete to return an error on not found. That coupled with relaxing the expected code to FailedPrecondition or NotFound would enable mongo to conform to the spec.

**Comment by jba** [comment]
@vangent @shantuo thoughts?

**Comment by jba** [comment]
One other piece of the puzzle: the reason Delete doesn't return an error is to make it easier to use in bulk actions. Currently, the spec of `ActionList.Do` says that it returns on first error, with the number of successfully completed actions. A Delete that failed if a document didn't exist would be annoying to use in bulk delete situations where some of the documents might happen to already be deleted. But maybe the spec of `Do` is too strong: as @clausti points out, we can't be sure that providers will execute action lists in any kind of order.

So maybe the right solution is for Delete to return an error for nonexistent documents, coupled with relaxing the semantics of `Do` to attempt all the actions, and return errors for each failed action.

**Comment by vangent** [comment]
Another option would be to optimistically try the write/delete; if it succeeds, great, we're done. If it fails, make another `Get` RPC with just the ID to determine which error to return.

**Comment by vangent** [comment]
+1 to having Delete return an error for nonexistent documents; the user can always choose to ignore it.

I don't quite follow the discussion for `Do`. What do the different providers do for errors during bulk actions?

**Comment by jba** [comment]
Good idea about only making the RPC in the error case. Did that.

Providers' treatment of bulk actions differs more than I thought. Firestore processes them atomically and in order; dynamoDB says it might do them in parallel; mongo supports both modes.

Since Firestore is atomic, you get either one error or none. It looks like mongo tells you exactly which actions failed. I believe that the way bulk operations work in dynamo, the individual actions can't fail. Puts and Deletes can't have preconditions, so they always succeed; Gets of nonexistent documents just don't return a value; and Updates are not allowed.

Bottom line: I think we should relax the semantics of action lists (renamed to action sets) to drop ordering, and return a list of errors. I'll start a PR for that tomorrow.

**Comment by vangent** [comment]
SGTM.

**Comment by shantuo** [comment]
+1 on relaxing the semantics.

Dynamo's batch operations are not atomic and returns the unprocessed items.

**Comment by jba** [comment]
I don't think we can use Dynamo's batch operations anyway. They don't allow conditions, which we need for revisions. We'll also need a condition on delete to make it return an error on not found.

**Comment by jba** [comment]
Thought about this idea overnight.

One of the motivations for sequential lists of actions is the ability to do certain small optimizations. For example, Dynamo's delete operation can return the document immediately before deletion, so
```
coll.Actions().Get(doc).Delete(doc).Do(ctx)
```
can be optimized to one RPC for Dynamo. But that assumes the actions are executed in order. 

Also, we have the ability to do ordered writes in a single RPC for both Mongo and Firestore; it seems a shame not to support that.

So I think we should keep ordered execution, but also support unordered via an option. 
- Since the bulk RPCs of Dynamo don't help us, we'd just run unordered actions concurrently.
- Mongo is easy: we'd just set the ordered option to false. 
- Firestore probably wouldn't change: we'd still do unordered actions as a single commit. We could explore running them concurrently to see if that was faster.

**Comment by shantuo** [comment]
The Transact API's of Dynamo supports conditions, but they are atomic.

**Comment by jba** [comment]
We should use those if/when possible. It's OK that they're atomic. The firestore implementation does basically the same thing.

**Comment by jba** [comment]
We've redone action lists to be unordered and independent, and relaxed the error semantics, so this is done.
